[MRG] Added API documentation chapter

mrbean-bremen opened the pull request:
- see #649

#### What does this implement/fix? Explain your changes.
The documentation created by this PR can be accessed [here](https://858-14006067-gh.circle-artifacts.com/0/home/ubuntu/pydicom/doc/_build/html/index.html).
I did some rough categorization of the modules in the API - improvements are welcome.

darcymason commented:
Thanks for digging into this, @mrbean-bremen.

A couple of quick thoughts
* maybe the Dataset section should go first?  I think it would be much more referenced than others
* when I view the output, right after DicomDir is a huge section of all the DicomDir (FileDataset/Dataset) methods and properties.  Is it possible to show only the extras of the subclass, and a link to the parent class for the inherited methods and properties? Seems ideal to me to keep most of the docs for those in one place (the main parent class) particularly when there is little added in the subclass.

mrbean-bremen commented:
Sure, `Dataset` at the beginning makes more sense. 
For `DicomDir`, I can just restrict the shown methods to the relevant (e.g. new) ones. Also, it may be better to show the base classes first - currently the classes are ordered alphabetically.

mrbean-bremen commented:
@darcymason - I did some re-ordering, removed the documentation of inherited members (they still show up in the methos overview after the class header), and added some sub-headings for each module to get some structure. Please [check it out](https://4-65673081-gh.circle-artifacts.com/0/home/ubuntu/pydicom/doc/_build/html/api_ref.html).

darcymason commented:
Looks really good!  I think it is almost ready to go, One little nitpick is that the examples for Dataset are not in a code block with one statement per line like in other examples sections.  Is that an easy fix?

mrbean-bremen commented:
Ah, didn't notice that! I think there are just some newlines missing (same in filereader), shall be easy to fix.
